firedrake/rf_eigensolver.py

```diff
@@ -0,0 +1,661 @@
+
+    def eigenfunction(self, i: int):
+        r"""Return the i-th eigenfunction of the solved problem.
+
+        Returns
+        -------
+        (Function, Function)
+            The real and imaginary parts of the eigenfunction.
+        """
+        if self._problem.restrict:
+            eigenmodes_real = Function(self._problem.restricted_space)
+            eigenmodes_imag = Function(self._problem.restricted_space)
+        else:
+            eigenmodes_real = Function(self._problem.output_space)
+            eigenmodes_imag = Function(self._problem.output_space)
+        with eigenmodes_real.dat.vec_wo as vr:
+            with eigenmodes_imag.dat.vec_wo as vi:
+                self.es.getEigenvector(i, vr, vi)
+        if self._problem.restrict:
+            eigenmodes_real = Function(self._problem.output_space).interpolate(eigenmodes_real)
+            eigenmodes_imag = Function(self._problem.output_space).interpolate(eigenmodes_imag)
+        return eigenmodes_real, eigenmodes_imag
+
+
+def new_restricted_control_variable(reduced_functional: ReducedFunctional, function_space: WithGeometry, dual: bool=False):
+    """Return new variables suitable for storing a control value or its dual
+        by interpolating into the space over which the 'ReducedFunctional' is
+        defined.
+
+    Parameters
+    ----------
+        reduced_functional: The `ReducedFunctional` whose
+        controls are to be copied.
+        function_space: Function space to which we restrict variables to.
+        dual: whether to return a dual type. If False then a primal type is returned.
+
+    Returns
+    -------
+        tuple[OverloadedType]: New variables suitable for storing a control value.
+    """
+    return tuple(
+        Function(function_space).interpolate(control.control)._ad_init_zero(dual=dual)
+        for control in reduced_functional.controls
+    )
+
+
+def interpolate_vars(variables: VarsInterpolate, function_space: WithGeometry):
+    """
+    Interpolates primal/dual variables to restricted/unrestricted function spaces.
+
+    Parameters
+    ----------
+        variables:
+            Variables that are to be interpolated into unrestricted/restricted primal/dual spaces.
+        restricted_space:
+            The function space where `PETSc.Vec` objects will live.
+
+    Returns
+    -------
+            The same variable/variables but interpolated into the necessary function space.
+    """
+    if isinstance(variables, (tuple, list)):
+        if isinstance(variables[0], Function):
+            return tuple(Function(function_space).interpolate(v) for v in variables)
+        else:
+            return tuple(
+                Cofunction(function_space.dual()).interpolate(v) for v in variables
+            )
+    elif isinstance(variables, Function):
+        return Function(function_space).interpolate(variables)
+    else:
+        return Cofunction(function_space.dual()).interpolate(variables)
```

Comment:
You can use `pyadjoint.Enlist` to handle whether we have a list or not. I think this is the right way to do it.
```suggestion
    from ufl.duals import is_primal
    variables = Enlist(variables)
    V = function_space if is_primal(variables[0]) else function_space.dual()
    interp_vars = [Function(V).interpolate(v) for v in variables)
    return variables.delist(interp_vars)
```